src/lib.rs

```diff
@@ -418,13 +431,23 @@ impl PyDAG {
     ) -> PyResult<usize> {
         let p_index = NodeIndex::new(parent);
         let c_index = NodeIndex::new(child);
-        let should_check_for_cycle =
-            must_check_for_cycle(self, p_index, c_index);
-        let state = Some(&mut self.cycle_state);
-        if should_check_for_cycle
-            && algo::has_path_connecting(&self.graph, c_index, p_index, state)
-        {
-            Err(DAGWouldCycle::py_err("Adding an edge would cycle"))
+        if self.check_cycle {
+            let should_check_for_cycle =
+                must_check_for_cycle(self, p_index, c_index);
```

Comment:
remind me what this does?

Reply:
`must_check_for_cycle` basically just looks if the parent node has parents, the child node has children, and that there is not already an edge from parent to child. If any of these are not true we can skip the cycle checking because there is no way to introduce a cycle.

Although, now that I think about it there is a potential bug by making this opt-in where you could add an edge without the cycle checking enabled that that introduces a cycle, then set `self.check_cycle=True` and never catch it because this will skip the cycle detection where there is an existing edge. I'm not sure how to best handle this, maybe run `is_directed_acyclic_graph()` in the setter if it goes from False to True?